Jbrowse2 tool

abretaud opened the pull request:
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

This has been standing too long on my hard drive: here's a wrapper for jbrowse2 :tada: :fireworks: !
Implemented the same way as Jbrowse (though there might be other cool things to do now that it's more embeddable). I mostly copied the jbrowse1 wrapper, and tried to adapt.
It mostly works, but it needs some clean up, there are some options that are still there but don't work yet (or should be removed), some new datatypes to add support to, ....

An interesting thing in jbrowse2 is that many display options can be easily changed from the jbrowse ui (while in jbrowse1 it required tweaking json), which means I'm not sure if we need to (re)implement them all in the form...

Also to test with `planemo server` I had to hack a little as planemo doesn't support byte-range headers. I can probably share my notes if there's interest

hexylena commented:
Love how much can be re-used for this, that's fantastic

hexylena commented on tools/jbrowse2/jbrowse2.xml:
```diff
@@ -350,14 +359,15 @@ cp $output.files_path/index.html $output;
 
     <!-- TODO genetic code not yet supported in jbrowse2 https://github.com/GMOD/jbrowse-components/issues/1765 -->
 
+    <!-- TODO show linear view by default? -->
```
+1

hexylena commented:
Interesting, if you use `config.json` in the root, next to index.html, then it'll by default open on to a working jbrowse in the first assembly. Too easy.

![image](https://github.com/galaxyproject/tools-iuc/assets/458683/c76b68cd-6f56-4e3b-b849-1e9ace214d6f)

abretaud commented:
Ok, it seems like this one is *finally* ready to be reviewed/merged :tada: :tada: :tada: 

There's still a linting error `Unknown datatype [hic] used in param element`, which should be fixed by https://github.com/galaxyproject/galaxy/pull/20874 --> it's not clear to me why the new datatype is not taken into account :sos: 

I think the features from [the tool that got merged in bgruening earlier](https://github.com/bgruening/galaxytools/tree/master/tools/jbrowse2) are present in this one, but overall looks much cleaner (e.g. cleaned up the useless code from jbrowse1 wrapper), and there are some additional features (ref name aliases, cytobands, customizable feature panel, metadata). Happy to make changes if anything is missing.

The tool supports linear genome view and synteny view too. And updating existing jbrowse instance works too, at least in most simple cases. I also added the possibility to reference data from remote urls: I know it's controversial, but I think it can be useful when reusing this tool in largest ebp-like projects, and the warning are quite visible about that.

I've left a short todo list in the README, but nothing blocking for a first merge I think.

Any reviews welcome!

ping @hexylena (if you're still interested :) ) @bgruening @cmdcolin

abretaud commented:
@Delphine-L would you have a VGP workflow and datasets to check if this tool is able to replace the bgruening one?

bgruening commented on tools/jbrowse2/jbrowse2.xml:
```diff
@@ -0,0 +1,1075 @@
+mkdir -p '$output.files_path/data';
+
+#if $action.action_select == "update":
+    cp -R $action.update_jbrowse.extra_files_path/data $output.files_path/;
```
quote

bgruening commented on tools/jbrowse2/jbrowse2.xml:
```diff
@@ -0,0 +1,1075 @@
+            </repeat>
+            <output name="output">
+                <extra_files name="galaxy.xml" type="file" value="out/01_all_tracks/galaxy.xml" compare="sim_size" />
+                <extra_files name="config.json" type="file" value="out/01_all_tracks/config.json" compare="sim_size" />
```
asserts?

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+            <yield/>
+        </requirements>
+    </xml>
+    <token name="@WRAPPER_VERSION@">galaxy0</token>
```
maybe just the 0 here? https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#tool-versions

name it version_suffix

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+    <token name="@WRAPPER_VERSION@">galaxy0</token>
+
+    <xml name="general_options">
+        <section name="jbgen" title="General JBrowse Options [Advanced]" expanded="false">
```
I personally don't like the [] - we don't use that anywhere else.

hexylena replied:
no, but it's also a bigger and more complex tool than 99%. Would () be better? We want to indicate that most folks probably don't need to look in there, unlike the selects that are more common options. Or prefixing with "Advanced options: " or something like that?

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+        <section name="jbstyle" title="Styling options [Advanced]" expanded="false">
+            <conditional name="track_style">
+                <param name="display" type="select" label="Display style" help="How the track will be displayed by default">
+                    <option value="LinearGCContentDisplay" selected="true">GC content</option>
```
only one option? Is that conditional needed then?

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+        <section name="jbstyle" title="Styling options [Advanced]" expanded="false">
+            <conditional name="track_style">
+                <param name="display" type="select" label="Display style" help="How the track will be displayed by default">
+                    <option value="LinearHicDisplay" selected="true">HiC</option>
```
only one option?

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+        <section name="jbstyle" title="Styling options [Advanced]" expanded="false">
+            <conditional name="track_style">
+                <param name="display" type="select" label="Display style" help="How the track will be displayed by default">
+                    <option value="LinearMafDisplay" selected="true">MAF</option>
```
only one option

hexylena replied:
I've only found the one MAF plugin @abretaud maybe you know of others? otherwise yeah, I can see @bgruening's argument :)

abretaud replied:
Yep there are many other plugins in https://jbrowse.org/jb2/plugin_store/, I have added to the TODO list (in readme) to see if we can add more in the tool (but later :smile: )
Also you can easily add custom plugins https://jbrowse.org/jb2/docs/config_guides/customizing_feature_colors/

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+            <expand macro="jexl_sanitize" />
+        </param>
+
+        <param label="Max height" name="max_height" type="integer" value="600" help="Maximum height that the track is permitted to reach in pixels."/>
```
add min (max) maybe?

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+    <xml name="track_metadata">
+        <section name="metadata" title="Track metadata [Advanced]" expanded="false">
+            <param name="galaxy_metadata" label="Show Galaxy track metadata" type="boolean" checked="true" truevalue="true" falsevalue="false" help="Make Galaxy metadata available in 'About track' window and faceted track selector" />
+            <param name="metadata_bonus" type="data" format="tsv" label="Additional track metadata" optional="true" help="Display additional track metadata from a TSV file" />
```
tabular or at least in addition

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+                </sanitizer>
+            </param>
+
+            <param label="Font size" name="font_size" type="integer" value="10" />
```
min/max

bgruening commented on tools/jbrowse2/macros.xml:
```diff
@@ -0,0 +1,355 @@
+            </param>
+            <param name="quaternary_color" type="color" label="Quaternary color" value="#ffb11d">
+                <sanitizer>
+                    <valid initial="string.ascii_letters,string.digits">
```
do we really need a sanitizer for type=color?

mvdbeek commented:
Attention: deployment skipped!

https://github.com/galaxyproject/tools-iuc/actions/runs/18189601014

abretaud commented:
Fixed now, thanks thanks thanks, quite happy to have it merged now!
